Replace "alpn" with "transport"/"no-default-transport"

bemasc opened the pull request:
This is an alternative to #89, based on our discussion in #73.

This is the "maximalist" approach, including a new IANA registry and a (backwards-compatible) update to RFC 7838.

Current text is definitely a rough cut.  We can clean up the requirements language if we agree this is the right direction.

DavidSchinazi commented:
In general I do like how this disconnects from Alt-Svc to address the problems mentioned by @davidben. However, I have two questions.

1) I'm not sure what the purpose of `no-transport` is. Having a definition of default transports for a scheme sounds like something that will become very rapidly obsolete. Why not simply just use `transport`?

2) How do we think this will interact with HTTP/3 or QUIC versions? In the TLS use-case, specifying `tls` is sufficient because TLS has an in-band version negotiation mechanism that does not cost round-trips, and ALPN also doesn't cost round trips. For QUIC, it's not quite certain we'll have that, so it might help to indicate the QUIC version? All that said, if we tighten this spec to have `quic` mean only QUICv1, then we can figure that out later.

dmcardle commented:
One thing to point out is that this causes the "port" parameter to have a dual meaning.   Suppose we build an HTTPSSVC record with `transport=quic transport=tls port=123` — the "port" parameter now applies to both UDP  and TCP!

MikeBishop commented:
> Having a definition of default transports for a scheme sounds like something that will become very rapidly obsolete.

It quite possibly will.  The concern that prompted this set of designs was that someone might publish a record indicating H3 support and providing ESNI keys, and thereby restrict those keys to the H3 endpoint.  It then makes it possible to disable ESNI over TLS/TCP by blocking the QUIC connection.

By requiring the record to explicitly disclaim H1/H2 support if it intends not to offer them on a given record, this accidental misconfiguration and associated attack become harder.

Specifically, I think that a list of transports is likely to have more staying power than particular versions.  Even if, however unlikely this seems, H1 becomes obsolete in the near future, I suspect HTTP over TLS will still exist in some negotiable shape.

davidben commented:
I don't think `no-transport` addresses those concerns. They still apply, with or without this set manipulation complexity. It merely adds additional complexity concerns. Clients, in particular, parse untrusted input, so needless complexity in formats must be dropped.

At the end of the day, if it is possible to express an ESNI key for the QUIC path without an ESNI key for the TCP path, the client needs to do *something* in response to it. The draft currently says the interpretation is that you disable the TCP path, which is unacceptable (see https://github.com/MikeBishop/dns-alt-svc/issues/73#issuecomment-565681979). The alternative non-rejecting interpretation is to silently lose ESNI, which is also unacceptable.

Thus it *must* be considered a misconfiguration. Well-designed mechanisms make misconfigurations impossible to spell. Example: advertising QUIC is purely additive, without a `no-transport` option, like the original `Alternate-Protocol` header. `Alt-Svc`-like designs regress this.

Alas, the reason for this regression was to support using different IPs for QUIC. We could do a two-level thing, with ESNI being specified at the top level, but this is perhaps a little goofy:
```
tcp.example.com; ipv4hint=1.2.3.4; port=443; esniconfigs=...; alt-transports={
  quic={quic.example.com; ipv4hint=5.6.7.8; port=444},
  fancyprotocol={fancyprotocol.example.com; ipv4hint=9.10.11.12; port=445},
}
```

Failing that, we are left with a misconfiguration that is otherwise syntactically valid. This is unstable under ecosystem effects. To counter this, the spec *must* define requirements to reject as loudly as feasible so administrators fix it. Here is a concrete possibility:
* There is a single "default" transport: whatever you would have done with A/AAAA records. It is specified by omitting the `transport` parameter and doesn't need a name.
* Remove the odd filtering and empty string rules around `esniconfig`. Optionally filtering an HTTPSSVC set complicates the rule below. If you want ESNI, you need to specify it on each HTTPSSVC arm. (A network attacker can always redirect you to the other machines, so having ESNI-capable and ESNI-incapable IPs that both speak for the public name is an inherently insecure configuration.)
* In a set of HTTPSSVC records, there must be one entry for the default transport. If there isn't, this is a misconfiguration:
  * Authoritative DNS servers reject this at config-time and do not advertise it at all.
  * Config file formats SHOULD make it hard to express the misconfiguration. Maybe add a `no-default-transport` token and compile it out? This is a local decision, so pick something.
  * If the authoritative DNS server is non-compliant, clients silently drop the record and fall back to A/AAAA. (Failing the HTTP request would be ideal but probably too much due to DNS caching.)

MikeBishop commented:
> Thus it _must_ be considered a misconfiguration. Well-designed mechanisms make misconfigurations impossible to spell.

And here, I think, is the distinction.  This isn't *necessarily* a misconfiguration, but it's potentially a misconfiguration.  The goal isn't to make H3-only impossible to say, it's to make it difficult to say accidentally.

Still, I think it's probably a fair assertion that HTTP services are practically guaranteed to include a TCP endpoint for the foreseeable future, and making that sound alarm bells if violated is a goal we can all agree on.  Perhaps requiring you to have a TCP entry, even if that entry then says `null` as the destination, is a good compromise?

davidben commented:
You quoted the conclusion without the rationale which led to it. The need for that to be a misconfiguration follows from the other two possible interpretations (drop TCP fallback or silently leak SNI) not working. A TCP entry with null destination would not work for browser scenarios, where site failures [impact users](https://www.w3.org/TR/html-design-principles/#priority-of-constituencies). I would push for Chrome to either ignore the spec and treat it as a misconfiguration regardless, or punt HTTPSSVC in favor of a more suitable record type.

Note that, without ESNI, HTTPSSVC already doesn't allow an h3-only configuration anyway. A/AAAA path is always an implicit route. It only accidentally happens for ESNI-capable clients as a quirk of ESNI's filtering bits.

bemasc commented:
Thanks for the comments.  I've uploaded a new version that tries to find a compromise between @MikeBishop and @davidben's suggestions.  I think the key line is:

> clients MAY reject the entire SVCB RRSet and fall back to basic connection establishment if all of the RRs indicate "no-default-transport"

This is similar to some of @davidben's suggestions.

While this doesn't make such RRSets non-compliant, it does clarify that clients are permitted to enforce such a requirement.  Assuming some popular clients adopt this behavior, such RRSets will be difficult to create accidentally, low-impact for end-users if admins do create them, and reasonably obvious to admins when testing or observing stats.

I haven't made any changes to the "esniconfig" logic, because I believe it is a separable question, and probably best addressed in a different PR.

PTAL!

dmcardle commented:
> One thing to point out is that this causes the "port" parameter to have a dual meaning. Suppose we build an HTTPSSVC record with `transport=quic transport=tls port=123` — the "port" parameter now applies to both UDP and TCP!

Any thoughts on this problem?

Quoting the introductory example from the latest patch:

```
    svc.example.net.  7200  IN HTTPSSVC 2 svc3.example.net. (
        additional-transport=quic port=8003 esniconfig="..." )
    svc.example.net.  7200  IN HTTPSSVC 3 svc2.example.net. (
                                        port=8002 esniconfig="..." )
```
Doesn't the first HTTPSSVC record imply "quic" and "tls" are both available over port 8003 on svc3.example.net?  How would you express a different port for QUIC?

bemasc commented:
> > One thing to point out is that this causes the "port" parameter to have a dual meaning. Suppose we build an HTTPSSVC record with `transport=quic transport=tls port=123` — the "port" parameter now applies to both UDP and TCP!
> 
> Any thoughts on this problem?

I updated the text to allude to this (line 162).

> Doesn't the first HTTPSSVC record imply "quic" and "tls" are both available over port 8003 on svc3.example.net?

Yes.  Sharing a port like this is a fairly common configuration (e.g. Google runs HTTPS over TLS and QUIC, both on port 443 on the same IPs).

> How would you express a different port for QUIC?

I would add `no-default-transport` to the QUIC RR, indicating that TLS is not supported.

enygren commented:
I am liking the direction of this, but this is a broad-reaching enough change that we should expand this proposal/discussion to include a larger audience when we're ready. A bunch of comments and questions:

    Can we name this something easier to write?
    The usability of "additional-transport" seems poor.
    Also, it is bikeshedding, but having a good name that
    these tokens and their registry can be referred
    to with and distinguished from ALPN seems critical.
    (So far "TP" is the best I've come up with as writing
    tp=tls is more readable.)

    I'm not sure all encodings / data types have
    an empty value as meaningful.

    We should talk about the relationship of {additional-transport /
    transport protocol} to ALPN. If we are going to go this path, this
    seems like a broader conversation to have that we should loop people
    like Martin Thompson and Tommy Pauly into as well. (There may be a
    question as to whether there is value to also supporting ALPN, but
    that may be unneeded complexity.) How this relates to Alt-Svc is
    something that should likely have discussion in the HTTP and QUIC
    (and perhaps TAPS?) working groups as well once there is a concrete
    proposal. A question may be whether this should get factored into
    its own document as this is a significant enough change to Alt-Svc.

    One way to frame the difference may be that:
        $TP (whatever we call it) defines a transport layer
        or transport+security layer protocol (eg, TCP+TLS, UDP+QUIC).
        Each of these provide a way to communicate over the network
        in a way that a secure connection can be negotiated and
        initiated.
        ALPN is an aggreement as part of the $TP negotiation
        to agree on the upper application-layer protocol
        (eg, HTTP-over-HTTP/1.1 vs HTTP-over-HTTP/2).

    Closely related to this is the relation of the "transport protocol
    set" with ALPN values. (ie, are there cases where the client
    needs to be aware which ALPN protocols within the transport
    protocol an endpoint supports prior to connecting?)

    Does it make sense that the transport registry is HTTPS-specific
    or should a single registry be shared across HTTPS
    and other protocols? I'd be inclined towards a single registry,
    especially given the uniqueness properites with ALPN.
    Having a bunch of registries all sharing a key space
    sounds like a disaster.

    There are legitimate use-cases for no-default-transport
    outside ot web browsers. It would be entirely reasonable
    for web services in a closed ecosystem to only support HTTP/3.

    Do we need to update something to indicate that
    the ALPN registry also doesn't collide with this?
    Does IANA have a precedent for this? Would merging
    these into a single registry with a new column (for if
    it is a $TP or an ALPN) be easier?

    We may want to be clear in our initial registry entries
    that these are "TLS over TCP" and "QUIC over UDP" ?

    We should provide guidance for how servers make
    a transition for Alt-Svc from alpn to these additional-transports.
    (ie, recommend that they include both, and we should probably
    show this in our example.)

    For "tls", what ALPN protocols does a server need to support?
    Could a server be HTTP/2 only or HTTP/1.1 only? I guess
    this will get figured out during the TLS handshake negotiation?
    (How this relates to Alt-Svc is also important.)

MikeBishop commented:
ALPN and Transport Protocol are different layers.  TLS and QUIC share a common property:  They are able to agree on an application protocol as part of their handshake.  Thus, our need for $TP is merely to point at the correct transport, and let its internal negotiation mechanisms handle the rest.  The open question is what happens if we’re pointing to a transport that does not have an internal protocol negotiation mechanism, but could nonetheless have several protocols (or incompatible protocol versions) on it.  I would suggest that’s no different from the world in TCP today, and the protocol has to do version agreement internal to itself.
 
However, there may also be cases like the QUIC version where having a hint in advance which version to pick would potentially speed up the process.  Those cases can just define a new parameter, I suppose.
 
As to whether it makes sense to require a particular ALPN value, I’d argue that it doesn’t.  ALPN is capable of restricting itself on either side – the client won’t offer anything it’s not willing to use, and the server won’t choose anything it’s not happy with.  The only value in knowing the “expected” outcome in advance is to then fail the connection if you don’t like what the server selected, even though you offered it as a valid choice.  (Or, potentially, for use with 0-RTT; but then you know in advance from the previous session without need of a SVCB record.)

bemasc commented on draft-ietf-dnsop-svcb-httpssvc.md:
```diff
@@ -334,7 +334,9 @@ multiple values.
 The presentation format for SvcFieldValue is a whitespace-separated
-list of the key=value pairs.  Each pair is presented in the following form:
+list of elements representing a key-value pair, with an absent value
+or "=" indicating an empty value.  Each element is presented in the
```
I added notes to the keys that don't support empty values.  Hopefully that should be enough to make it clear.

I made this adjustment because `transport=quic no-default-transport` seemed a lot clearer than `transport=quic no-default-transport=""`.  In general, I think its worth investing in clarity for zone file editors, since zone files get edited a lot more often than DNS software.

bemasc commented on draft-ietf-dnsop-svcb-httpssvc.md:
```diff
@@ -682,30 +685,37 @@ responses to the address queries that were issued in parallel.
+that transport.
+
+Clients SHOULD NOT attempt connection to a service endpoint whose
+transport set does not contain any compatible protocols.  To ensure
```
Done.

bemasc commented:
(Skipping questions addressed by Mike's comment and reordering a few things...)

> Can we name this something easier to write?
> The usability of "additional-transport" seems poor.

OK, I changed it to "transport".  I think that's short enough, and a lot more human-readable than "tp".

> Do we need to update something to indicate that
> the ALPN registry also doesn't collide with this?

I decided to try to solve this by taking a different approach.  In the latest commit, this draft now defines a new fake ALPN ("any") to indicate that ALPN should be negotiated.  This makes the HTTPS-Transport and ALPN registries completely independent (not sharing a keyspace, no need for conflict avoidance).

> If we are going to go this path, this seems like a broader conversation to have...

Whatever we do, it'll definitely become a bigger conversation, but I'd like to rev the draft to include our best consensus idea before soliciting input from other WGs.

> Does it make sense that the transport registry is HTTPS-specific
> or should a single registry be shared across HTTPS
> and other protocols? I'd be inclined towards a single registry,
> especially given the uniqueness properites with ALPN.
> Having a bunch of registries all sharing a key space
> sounds like a disaster.

OK, I've changed the design to avoid that disaster.  I do think it makes sense to have separate transport registries.  If we have a single transport registry that describes the allowable transports for HTTP, HTTPS, SSH, and DNS, it's going to be a real mess to explain which ones you can use when.

> There are legitimate use-cases for no-default-transport
> outside ot web browsers. It would be entirely reasonable
> for web services in a closed ecosystem to only support HTTP/3.

The relevant text is:
```
For compatibility with clients that require default transports,
zone operators SHOULD ensure that at least one RR in each RRSet supports the
default transports.
```
I think it's clear enough that one is free to violate that SHOULD in a closed ecosystem.

> We may want to be clear in our initial registry entries
> that these are "TLS over TCP" and "QUIC over UDP" ?

The "TLS" line is referenced to RFC 2818, which seems like it ought to be the place for that to be specified.  I admit that RFC 2818 could be a little clearer about that, but personally I don't think it's ambiguous.  Hopefully the HTTP-over-QUIC RFC will be clearer.  If you think it's unclear, I guess we could add a "meaning" column to the registry.

> We should provide guidance for how servers make
> a transition for Alt-Svc from alpn to these additional-transports.
> (ie, recommend that they include both,

Done

> and we should probably show this in our example.)

I didn't add an example of this, but we can if you want.  It gets us back into some thorny territory with esniconfig...

enygren commented on draft-ietf-dnsop-svcb-httpssvc.md:
```diff
@@ -139,12 +139,11 @@ Consider a simple zone of the form
-supports HTTPS, and the key=value pairs indicate that it prefers HTTP/3
-but also supports HTTP/2.  The records can also include other information
+The presence of this record indicates to clients that simple.example
+supports HTTPS, and the key=value pairs indicate that it supports QUIC
```
Perhaps "supports HTTPS-over-QUIC (HTTP/3) in addition to HTTPS over TLS" ?

enygren replied:
Adding a "meaning" column to the registry may make sense.

It seems like we're making enough changes to Alt-Svc here that we may need to split some of these out to an alt-svc-bis draft?   Switching Alt-Svc to use ALPN=any but focusing on transport may make sense.  At a minimum we likely should include a section here on how filtering works for Alt-Svc in a world where records are like:

       Alt-Svc: any="pool.service.example:443" transport=quic, any="pool.service.example:443" transport=tls 

As part of this change we should include a reminder that clients using HTTPSSVC or AltSvc with "any" MUST still send TLS SNI (or ESNI when esniconfig is specified).

Going back to a question from @MikeBishop, do we still want a hinting/filtering mechanism for ALPN?  This can be useful for letting clients using an experimental protocol know that it is available on a pool with a different name.  That has been one Alt-Svc use-case for ALPN.  Or does this add too much complexity?

bemasc commented:
This has been rebased and substantially simplified. There is no longer any need for the "any" ALPN and other interaction with Alt-Svc.  PTAL.

davidben commented on draft-ietf-dnsop-svcb-httpssvc.md:
```diff
@@ -743,38 +745,49 @@ responses to the address queries that were issued in parallel.
+sequence of octets is a permissible value.  This key MAY appear multiple times
+with different values.
+
+The value of "no-default-transport" MUST be empty, and clients SHOULD reject
```
s/SHOULD/MUST/

It's a syntax error. Syntax errors are fatal.

bemasc replied:
Done

bemasc commented on draft-ietf-dnsop-svcb-httpssvc.md:
```diff
@@ -743,38 +745,49 @@ responses to the address queries that were issued in parallel.
+"no-default-transport" SvcParamKey is present.  The presence of a value in
+the transport set indicates that this service endpoint, described by
+SvcDomainName and the other parameters (e.g. "port") offers service with
+that transport.
```
> This design more-or-less forces all new UDP-based protocols for transporting HTTP to share protocol invariants with QUIC.

I'm not quite sure what this means, or if it's true.  There may indeed be two UDP transports that can't share a port.  That's fine; they can just go in separate RRs with different port numbers.

We could certainly restructure this to require one transport per RR, or to express multiple ports per RR (TCP and UDP or per-transport), but the former seems inefficient, and the latter seems overcomplicated.  Sharing a port number between TCP and UDP may not be precisely principled, but in practice IANA [almost always](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers) assigns both ports to the same service, so I think it's fairly intuitive.

As you noted, sharing an RR is mostly a space optimization, especially with ESNIConfig.  My guess is that the majority of IPs that have QUIC and TLS with the same ESNIConfig will also have them at the same port number, so one-port-number-per-RR gets us most of the space savings with the least complexity.

davidben replied:
Sorry, I should have elaborated on "more-or-less". Encoding separate RRs for port purposes in this proposal is fussy (you need a no-default-transport) and costs a duplicate ESNI config. (Which we presumably care about or we'd go with the flattened design.) That means there will be pressure at the protocol design phase to share invariants, which means we effectively are saying this.

That's not to say this is fatal. Maybe each scheme only needs two sets of protocol invariants? QUIC folks have no doubt thought about this more than me so I'd like to know what they think. My main position is that, if we're strongly incentivizing this, we need to recognize that and decide we're okay with it, and not hide behind the verbose alternative existing.

bemasc replied:
> Encoding separate RRs for port purposes in this proposal is fussy (you need a no-default-transport) and costs a duplicate ESNI config. (Which we presumably care about or we'd go with the flattened design.)

Agreed, although this syntax was also motivated by simplicity for zone owners when QUIC and TLS share all parameters, which we took to be the common case.

> That means there will be pressure at the protocol design phase to share invariants, which means we effectively are saying this.

I think this is at most a weak incentive, not a significant pressure on future designs.  Suppose NOTQUIC is a secure transport protocol over UDP that cannot share a port with QUIC.  (I think this is an unlikely scenario but we should nonetheless support it.)  In the current design, this would be written as
```
... port=1234 transport=quic esniconfig="abc..." ; TLS and QUIC
... port=4321 transport=notquic no-default-transport esniconfig="def..." ; NOTQUIC
```
In a flattened design, this would instead be
```
... port=1234 esniconfig="abc..." ; TLS
... port=1234 transport=quic esniconfig="abc..." ; QUIC
... port=4321 transport=notquic esniconfig="def..." ; NOTQUIC
```
I don't see a major advantage in comprehensibility between these options.  On balance, I prefer the former; duplicating the ESNIConfig seems to invite typos and misconfigurations that will not easily be caught.

BTW, even if NOTQUIC can't share a UDP port with QUIC, users can still avoid `no-default-transport` by running TLS on the TCP port of the same number.  This seems like a reasonable, perhaps likely configuration, for the same reasons that it appears to be popular with QUIC.

davidben replied:
The problem with the half-flattened scheme is this design is extremely inconsistent in its principles. It simultaneously believes `no-default-transport` is a rare misfeature only to be used for questionable deployment patterns, at the same time it is a pattern for new protocols. It then simultaneously believes duplicating the ESNIConfig is a problem and should be avoided, yet recommends it as a pattern for new protocols. All this mixup then pressures new protocols to share UDP ports to avoid all these design flaws, yet it also insists this pressure doesn't exist.

> BTW, even if NOTQUIC can't share a UDP port with QUIC, users can still avoid no-default-transport by running TLS on the TCP port of the same number. This seems like a reasonable, perhaps likely configuration, for the same reasons that it appears to be popular with QUIC.

No, this is not a reasonable or likely configuration. Moreover, it's dangerous because it further incentivizes existing TCP HTTP servers to not check port numbers against the HTTP `Host` header. (This check is a security requirement for sites running different origins on different ports as it's the only thing preventing the network from mixing ports up.)

It's popular with QUIC because, barring any reason to pick another port, 443 is the natural one in order to match with the origin. HTTPSSVC having broken all these standard assumption is not going to suddenly make people want to deploy TCP over the corresponding port for NOTQUIC. That doesn't really make any sense at all.

bemasc replied:
It sounds like you're proposing to drop the requirement that zone authors need to take an explicit action to disable TLS support, as in your flattened example in this thread.  If you're willing to drop that requirement, I agree that we can find a more principled design.

Personally, I would lean toward more multiplexing, e.g.
```
... ports=tls:1234,quic:1234 esniconfig="abc..."
... ports=notquic:4321 esniconfig="def..."
```

enygren replied:
I'm not a fan of multiplexing based on port and making it primary in the model.  There may be transports in the future that move away from numeric ports or have a different model that isn't port-based.  As an example, if we had a future L4 protocol assignment for some QUIC successor (rather than requiring it to be over UDP).

bemasc replied:
I think extension to portless protocols would work fine in this model.  It would just require defining one or more new SvcParamKeys.  For example, we could have
```
... ports=tls:443,quic:443,quic:8443 l4quic esniconfig="abc..."
```
where `l4quic` is a new empty-valued SvcParamKey indicating support for the (hypothetical) QUIC-at-L4 protocol.

Basically, ServiceForm is extensible enough that we can probably convey whatever information we need to in the future.

davidben replied:
> It sounds like you're proposing to drop the requirement that zone authors need to take an explicit action to disable TLS support, as in your flattened example in this thread.

I mean, I never requested that property to begin with. :-) I think something got lost in translation here.

The deployability requirement is that is *impossible* to disable the default transport. Possible with explicit action doesn't solve anything. It should either be syntactically impossible (preferred) or, if that can't be avoided, it needs to be easy to describe and detect (that's why I've been trying to avoid giving it a name that differs by scheme), with the expectation that authoritative servers reject it at the config level and clients reject it if they see it.

From there, it would be nice if the wrong thing was easy to avoid, but if authoritatives perform the mandatory checks, that helps smooth things over a bit.

Let's sync after the long weekend. GitHub PRs aren't a great medium for design discussions. I think we need something more synchronous and high-bandwidth. This DNS record has gotten far too complicated. We need to cut down the flexibility by an order of magnitude before this is viable.

enygren commented on draft-ietf-dnsop-svcb-httpssvc.md:
```diff
@@ -743,38 +745,49 @@ responses to the address queries that were issued in parallel.
+in an RR.  For compatibility with clients that require default transports,
+zone operators SHOULD ensure that at least one RR in each RRSet supports the
+default transports.
 
```
Perhaps add:
"The wire format for the value of no-default-transport is empty (a SvcParamValue with a length of zero octets)."
or something else to be crystal clear that value is still present in the wire format but empty.

bemasc replied:
OK, I added clarifying text to this effect.

bemasc commented:
It sounds like we have a lot more to discuss in the finer details of this arrangement, but it's clear that this PR is closer to our consensus than what's in `master`, so I'm going merge this as-is.  We can improve the representation in followup PRs.
